cranelift/codegen/src/isa/x64/lower.rs

```diff
@@ -870,7 +999,65 @@ fn lower_insn_to_regs<C: LowerCtx<I = Inst>>(
 
         Opcode::Trap | Opcode::ResumableTrap => {
             let trap_info = (ctx.srcloc(insn), inst_trapcode(ctx.data(insn)).unwrap());
-            ctx.emit(Inst::Ud2 { trap_info })
+            ctx.emit_safepoint(Inst::Ud2 { trap_info });
+        }
+
+        Opcode::Trapif | Opcode::Trapff => {
+            let srcloc = ctx.srcloc(insn);
+            let trap_code = inst_trapcode(ctx.data(insn)).unwrap();
+
+            let cc = if matches_input(ctx, inputs[0], Opcode::IaddIfcout).is_some() {
+                let condcode = inst_condcode(ctx.data(insn));
+                // The flags must not have been clobbered by any other instruction between the
+                // iadd_ifcout and this instruction, as verified by the CLIF validator; so we can
+                // simply use the flags here.
+                CC::from_intcc(condcode)
+            } else if op == Opcode::Trapif {
+                let condcode = inst_condcode(ctx.data(insn));
+                let cc = CC::from_intcc(condcode);
+
+                // Verification ensures that the input is always a single-def ifcmp.
+                let ifcmp_insn = matches_input(ctx, inputs[0], Opcode::Ifcmp).unwrap();
+                emit_cmp(ctx, ifcmp_insn);
+                cc
+            } else {
+                let condcode = inst_fp_condcode(ctx.data(insn)).unwrap();
+                let cc = CC::from_floatcc(condcode);
+
+                // Verification ensures that the input is always a single-def ffcmp.
+                let ffcmp_insn = matches_input(ctx, inputs[0], Opcode::Ffcmp).unwrap();
+                {
+                    // The only valid CC constructed with `from_floatcc` can be put in the flag
+                    // register with a direct float comparison; do this here.
+                    let input_ty = ctx.input_ty(ffcmp_insn, 0);
```

Comment:
I wasn't sure about this, to be honest: there aren't many opportunities for reuse, since the actual Fcmp lowering does things quite differently. Ok on principle, though.